File: mpi-1x-hypack-2013/point-to-point-mpi-codes/f77-lang-pt2pt-mpi/pie_collective.f
c
c********************************************************************
c
c	C-DAC Tech Workshop : hyPACK-2013
c               October 15-18, 2013
c
c   Example 11		: pie_collective.f
c
c   Objective           : To compute the value of PI by numerical integration.
c                         MPI Collective communication library calls are used. 
c                         This example demonstrates the use of
c
c                         MPI_Init
c                         MPI_Comm_rank
c                         MPI_Comm_size
c                         MPI_Bcast
c                         MPI_Reduce
c                         MPI_Finalize
c
c   Input               : The number of intervals.
c
c   Output              : The calculated value of PI.
c
c   Necessary Condition : Number of Processes should be 
c                         less than or equal to 8.
c
c   Created             : August-2013
c
c   E-mail              : hpcfte@cdac.in     
c
c
c*********************************************************************

      program main

      include 'mpif.h'

      double precision  PI25DT
      parameter        (PI25DT = 3.141592653589793238462643d0)

      double precision  mypi, pi, h, sum, x, f, a
      integer NoInterval, MyRank, Numprocs, interval, Root

      
C     function to integrate

      f(a) = 4.0d0 / (1.0d0 + a*a)

      call MPI_INIT( ierr )
      call MPI_COMM_RANK( MPI_COMM_WORLD, MyRank, ierr )
      call MPI_COMM_SIZE( MPI_COMM_WORLD, Numprocs, ierr )

      Root = 0 
     
 10   if ( MyRank .eq. Root ) then
         write(6,98)
 98      format('Enter the number of intervals: (0 quits)')
         read(5,99) NoInterval
 99      format(i10)
      endif
      
      call MPI_BCAST(NoInterval,1,MPI_INTEGER,0,MPI_COMM_WORLD,ierr)

c     check for quit signal
      if ( NoInterval .LE. 0 ) then
        if(MyRank .eq. Root) print*,"Invalid Number of Intervals"
	goto 30
      endif

C     calculate the interval size

      h = 1.0d0/NoInterval
      sum  = 0.0d0

      do 20 interval = MyRank+1, NoInterval, Numprocs
         x = h * (dble(interval) - 0.5d0)
         sum = sum + f(x)
 20   continue
      mypi = h * sum

C     collect all the partial sums
      call MPI_REDUCE(mypi,pi,1,MPI_DOUBLE_PRECISION,MPI_SUM,Root,
     $     MPI_COMM_WORLD,ierr)

C     node 0 prints the answer.
      if (MyRank .eq. Root) then
         write(6, 97) pi, abs(pi - PI25DT)
 97      format('  pi is approximately: ', F18.16,
     +          '  Error is: ', F18.16)
      endif

      goto 10

 30   call MPI_FINALIZE(ierr)
      stop
      end

File: mpi-1x-hypack-2013/collective-comm-comp-codes/c-lang-collect-mpi/pie-collective.c

/*
**********************************************************************
		C-DAC Tech Workshop : hyPACK-2013
                           October 15-18, 2013

  Example 2.8	    : pie_collective.c

  Objective           : To compute the value of PI by numerical integration.
                        MPI Collective communication library calls are used.

                        This example demonstrates the use of 
                        MPI_Init
                        MPI_Comm_rank
                        MPI_Comm_size
                        MPI_Bcast
                        MPI_Reduce
                        MPI_Finalize

  Input               : The number of intervals.

  Output              : The calculated value of PI.

  Necessary Condition : Number of Processes should be
                        less than or equal to 8.

   Created             : August-2013

   E-mail              : hpcfte@cdac.in     

**********************************************************************
*/


#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <mpi.h>

double func(double x)
{
    return (4.0 / (1.0 + x*x));
}

int main(int argc,char *argv[])
{
    int    NoInterval, interval;
    int    MyRank, Numprocs, Root ; 
    double mypi, pi, h, sum, x;
    double PI25DT = 3.141592653589793238462643;

    /*....MPI initialisation....*/
    MPI_Init(&argc,&argv);
    MPI_Comm_size(MPI_COMM_WORLD,&Numprocs);
    MPI_Comm_rank(MPI_COMM_WORLD,&MyRank);

    if(MyRank == Root){
	puts("\nEnter the number of intervals : ");
	scanf("%d",&NoInterval);
    }
        
   /*....Broadcast the number of subintervals to each processor....*/ 
   MPI_Bcast(&NoInterval, 1, MPI_INT, 0, MPI_COMM_WORLD);

   if(NoInterval <= 0){
      if(MyRank == Root) 
	 printf("Invalid Value for Number of Intervals .....\n");
      MPI_Finalize();
      exit(-1);
   }

   h   = 1.0 / (double)NoInterval;
   sum = 0.0;
   for(interval = MyRank + 1; interval <= NoInterval; interval += Numprocs){
       x = h * ((double)interval - 0.5);
       sum += func(x);
   }
   mypi = h * sum;

   /*....Collect the areas calculated in P0....*/ 
   MPI_Reduce(&mypi, &pi, 1, MPI_DOUBLE, MPI_SUM, Root, MPI_COMM_WORLD);

   if(MyRank == Root){
      printf("pi is approximately %.16f, Error is %.16f\n", 
						   pi, fabs(pi - PI25DT));
   }

   MPI_Finalize();

}




